include/picongpu/simulationControl/AdjustDomainSize.hpp

```diff
@@ -0,0 +1,363 @@
+            MPI_CHECK( MPI_Allgather(
+                &mpiPos,
+                1,
+                MPI_INT,
+                mpiPositions.data(),
+                1,
+                MPI_INT,
+                gc.getCommunicator().getMPIComm()
+            ));
+
+            // gather local sizes in the direction we are checking
+            std::vector< uint64_t > allLocalSizes( numMpiRanks );
+            uint64_t lSize = static_cast< uint64_t >( m_localDomainSize[ dim ] );
+            MPI_CHECK( MPI_Allgather(
+                &lSize,
+                1,
+                MPI_UINT64_T,
+                allLocalSizes.data(),
+                1,
+                MPI_UINT64_T,
+                gc.getCommunicator().getMPIComm()
+            ));
+
+            uint64_t offset = 0;
+            for( size_t i = 0; i < mpiPositions.size(); ++i )
+            {
+                if( mpiPositions[ i ] < mpiPos )
+                    offset += allLocalSizes[ i ];
+            }
+
+            /* since we are not doing independent reduces per slice we need
+             * to adjust the offset result by dividing with the number of
+             * MPI ranks in all other dimensions.
+             */
+            offset /= static_cast< uint64_t >( m_numDevices.productOfComponents() / m_numDevices[ dim ] );
+            m_localDomainOffset[ dim ] = static_cast< int >( offset );
+
+        }
+
+        /** ensure that the local size is a multiple of the supercell size
+         *
+         * @param dim dimension to update
+         */
+        void multipleOfSuperCell( size_t const dim )
+        {
+            int const sCellSize = SuperCellSize::toRT()[ dim ];
+            // round up to full supercells
+            int const validLocalSize =
+                ( ( m_localDomainSize[ dim ] + sCellSize - 1 ) / sCellSize ) *
+                sCellSize;
+
+            if( validLocalSize != m_localDomainSize[ dim ] )
+            {
+                std::cout << "Dimension " << dim <<
+                    ": local grid size is not a multiple of supercell size. Adjust from " <<
+                    m_localDomainSize[ dim ] << " to " <<  validLocalSize << std::endl;
+                sizeChanged = true;
+            }
+
+            m_localDomainSize[ dim ] = validLocalSize;
+        }
+
+        /** ensure that we have a CORE and BORDER region
+         *
+         * Each region must have the size of at least one supercell.
+         *
+         * @param dim dimension to update
+         */
+        void minThreeSuperCells( size_t dim )
+        {
+            int superCellCount = m_localDomainSize[ dim ] / SuperCellSize::toRT()[ dim ];
+            if( superCellCount < 3 )
+                superCellCount = 3;
+
+            int validLocalSize = superCellCount * SuperCellSize::toRT()[ dim ];
+            if( validLocalSize != m_localDomainSize[ dim ] )
+            {
+                std::cout << "Dimension " << dim <<
```

Comment:
I guess for a user it is more natural to operate and see log messages in terms of x, y, z, not "dimension 0 / 1 / 2". Also in case, e.g. the output only has "dimension 1", without looking to the code it is not even clear if it means x or y.